File: src/tasks/cli/command/blockers/BlockersHandler.java
package tasks.cli.command.blockers;

import static java.util.Objects.requireNonNull;
import static tasks.cli.handlers.HandlerUtil.stringifyIfPopulated;
import static tasks.cli.handlers.HandlerUtil.verifyTasksAreMutuallyExclusive;

import io.reactivex.Completable;
import io.reactivex.Single;
import omnia.algorithm.SetAlgorithms;
import omnia.cli.out.Output;
import omnia.data.cache.Memoized;
import omnia.data.structure.Set;
import omnia.data.structure.tuple.Couple;
import omnia.data.structure.tuple.Tuple;
import tasks.cli.handlers.ArgumentHandler;
import tasks.cli.handlers.HandlerException;
import tasks.model.Task;
import tasks.model.TaskStore;

/** Business logic for the Blockers command. */
public final class BlockersHandler implements ArgumentHandler<BlockersArguments> {
  private final Memoized<TaskStore> taskStore;

  public BlockersHandler(Memoized<TaskStore> taskStore) {
    this.taskStore = requireNonNull(taskStore);
  }

  @Override
  public Single<Output> handle(BlockersArguments arguments) {
    /*
     * Ensure task isn't connected to itself.
     * This is a short-circuit, but is not strictly required because we still need to check if the
     * task graph is cyclical.
     */
    if (arguments.blockingTasksToAdd().contains(arguments.targetTask())) {
      throw new HandlerException("target task cannot block or be blocked by itself");
    }

    verifyTasksAreMutuallyExclusive(
        "ambiguous operation: blockers both added and removed: ",
        arguments.blockingTasksToAdd(),
        arguments.blockingTasksToRemove());

    return mutateAndProduceBeforeAfterSnapshot(arguments)
        .map(BlockersHandler::stringifyResults)
        .map(
            results ->
                Output.builder()
                    .appendLine(arguments.targetTask().render())
                    .appendLine(results)
                    .build());
  }

  private Single<Couple<Set<Task>, Set<Task>>> mutateAndProduceBeforeAfterSnapshot(
      BlockersArguments arguments) {
    return getTasksBlocking(arguments.targetTask())
        .flatMap(
            blockingTasksBeforeMutation ->
                mutateTaskStore(arguments)
                    .andThen(getTasksBlocking(arguments.targetTask()))
                    .map(blockingTasksAfterMutation ->
                        Tuple.of(blockingTasksBeforeMutation, blockingTasksAfterMutation)));
  }

  private Completable mutateTaskStore(BlockersArguments arguments) {
    return taskStore.value().mutateTask(
        arguments.targetTask(),
        mutator -> {
          if (arguments.clearAllBlockers()) {
            mutator.setBlockingTasks(arguments.blockingTasksToAdd());
          } else {
            arguments.blockingTasksToRemove().forEach(mutator::removeBlockingTask);
            arguments.blockingTasksToAdd().forEach(mutator::addBlockingTask);
          }
          return mutator;
        })
        .andThen(taskStore.value().writeToDisk());
  }

  private static Single<Set<Task>> getTasksBlocking(Task task) {
    return task.query().tasksBlockingThis().firstOrError();
  }

  private static Output stringifyResults(Couple<Set<Task>, Set<Task>> beforeAfterSnapshots) {
    return Output.builder()
        .appendLine(stringifyIfPopulated("current blockers:", beforeAfterSnapshots.second()))
        .appendLine(
            stringifyIfPopulated(
                "removed blockers:",
                getRemovedBlockers(beforeAfterSnapshots.first(), beforeAfterSnapshots.second())))
        .build();
  }

  private static Set<Task> getRemovedBlockers(Set<Task> before, Set<Task> after) {
    return SetAlgorithms.differenceBetween(before, after);
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/blockers/BlockersHandler.kt
package me.strangepan.tasks.cli.feature.blockers

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import java.util.function.Consumer
import java.util.function.Function
import omnia.cli.out.Output
import omnia.cli.out.Output.Companion.empty
import omnia.data.cache.Memoized
import omnia.data.stream.Collectors.toImmutableSet
import omnia.data.structure.Set
import omnia.data.structure.immutable.ImmutableSet
import omnia.data.structure.tuple.Couplet
import omnia.data.structure.tuple.Triple
import omnia.data.structure.tuple.Tuple
import omnia.data.structure.tuple.Tuplet
import me.strangepan.tasks.cli.command.common.CommonArguments
import me.strangepan.tasks.cli.handler.ArgumentHandler
import me.strangepan.tasks.cli.handler.HandlerUtil
import me.strangepan.tasks.cli.handler.HandlerUtil.stringifyIfPopulated
import me.strangepan.tasks.engine.model.ObservableTaskStore
import me.strangepan.tasks.engine.model.Task
import me.strangepan.tasks.engine.model.TaskStore

/** Business logic for the Blockers command.  */
class BlockersHandler(private val taskStore: Memoized<out ObservableTaskStore>) : ArgumentHandler<BlockersArguments> {

  override fun handle(arguments: CommonArguments<out BlockersArguments>): Single<Output> {
    /*
     * Ensure task isn't connected to itself.
     * This is a short-circuit, but is not strictly required because we still need to check if the
     * task graph is cyclical.
     */
    HandlerUtil.verifyTasksAreMutuallyExclusive(
        "me.strangepan.tasks.engine.tasks cannot block or be blocked by themselves:",
        arguments.specificArguments().blockingTasksToAdd(),
        arguments.specificArguments().targetTasks())
    HandlerUtil.verifyTasksAreMutuallyExclusive(
        "ambiguous operation: blockers both added and removed: ",
        arguments.specificArguments().blockingTasksToAdd(),
        arguments.specificArguments().blockingTasksToRemove())
    return if (!arguments.specificArguments().blockingTasksToAdd().isPopulated
        && !arguments.specificArguments().blockingTasksToRemove().isPopulated
        && !arguments.specificArguments().clearAllBlockers()) {
      Single.just(empty())
    } else mutateAndProduceBeforeAfterSnapshots(arguments.specificArguments())
        .map(::stringifyResults)
        .collect(Output::builder, Output.Builder::appendLine)
        .map(Output.Builder::build)
  }

  private fun mutateAndProduceBeforeAfterSnapshots(arguments: BlockersArguments):
      Observable<out Triple<Task, Set<out Task>, Set<out Task>>> {
    return mutateTaskStore(arguments)
        .flatMapObservable { couplet ->
          Observable.fromIterable(arguments.targetTasks())
              .map { task ->
                val blockingTasksBeforeAfter = couplet.map(Function { store ->
                  store.lookUpById(task.id())
                      .map(Task::blockingTasks)
                      .orElse(ImmutableSet.empty())
                })
                Tuple.of(task, blockingTasksBeforeAfter.first(), blockingTasksBeforeAfter.second())
              }
        }
  }

  private fun mutateTaskStore(arguments: BlockersArguments): Single<Couplet<TaskStore>> {
    return Observable.fromIterable(arguments.targetTasks())
        .flatMapSingle { task ->
          taskStore.value().mutateTask(task) { mutator ->
            if (arguments.clearAllBlockers()) {
              mutator.setBlockingTasks(arguments.blockingTasksToAdd())
            } else {
              arguments.blockingTasksToRemove().forEach(Consumer(mutator::removeBlockingTask))
              arguments.blockingTasksToAdd().forEach(Consumer(mutator::addBlockingTask))
            }
            mutator
          }
        }
        .map { it.dropThird() }
        .reduce { firstCouple, secondCouple -> firstCouple.mapSecond { secondCouple.second() }}
        .toSingle()
        .map(Tuplet.Companion::copyOf)
  }

  companion object {
    private fun stringifyResults(
        beforeAfterSnapshots: Triple<out Task, out Set<out Task>, out Set<out Task>>): Output {
      return Output.builder()
          .appendLine(beforeAfterSnapshots.first().render())
          .appendLine(stringifyIfPopulated("current blockers:", beforeAfterSnapshots.third()))
          .appendLine(
              stringifyIfPopulated(
                  "removed blockers:",
                  getRemovedBlockers(beforeAfterSnapshots.second(), beforeAfterSnapshots.third())))
          .build()
    }

    private fun getRemovedBlockers(before: Set<out Task>, after: Set<out Task>): Set<out Task> {
      val afterIds = after.stream().map(Task::id).collect(toImmutableSet())
      return before.stream().filter { !afterIds.contains(it.id()) }.collect(toImmutableSet())
    }
  }
}